controllers/humiocluster_controller.go

```diff
@@ -2074,10 +2079,170 @@ func (r *HumioClusterReconciler) ensurePodsExist(ctx context.Context, hc *humiov
+//  2. Computes the zone from which the pod will be removed base on the current node allocation
+//  3. Iterates through pods and for the first one found in the specified zone, sends an eviction request to the node
+//  4. Checks if the eviction has started (with a timeout of 10 seconds)
+//  5. If the eviction has started, it periodically checks every 60 seconds if the eviction has been completed
+//  6. When the eviction is completed and there is no more data on that node, the node is unregistered from the cluster, and the pod is removed.
+func (r *HumioClusterReconciler) removePodGracefully(ctx context.Context, hc *humiov1alpha1.HumioCluster, podsInNodePool []corev1.Pod, req ctrl.Request) error {
+	// GetCluster gql query returns node ID and Zone
+	clusterConfig, err := helpers.NewCluster(ctx, r, hc.Name, "", hc.Namespace, helpers.UseCertManager(), true, false)
+	if err != nil {
+		return r.logErrorAndReturn(err, fmt.Sprintf("could not create a cluster config for the http client."))
+	}
+	humioHttpClient := r.HumioClient.GetHumioHttpClient(clusterConfig.Config(), req)
+	cluster, err := r.HumioClient.GetCluster(ctx, humioHttpClient, req)
+	if err != nil {
+		return r.logErrorAndReturn(err, "failed to get humio cluster through the GraphQL API")
+	}
+	getCluster := cluster.GetCluster()
+	nodeIdToPodMap := r.matchPodsToHosts(podsInNodePool, getCluster.GetNodes())
+
+	// Check Node Zones and gets the one with the most nodes. In case of a tie, the first zone is used
+	podRemovalZone, err := r.getZoneForPodRemoval(ctx, podsInNodePool)
+	if err != nil {
+		return r.logErrorAndReturn(err, "failed to get pod removal zone")
+	}
+	for _, pod := range podsInNodePool {
+		if pod.Labels[kubernetes.TopologyKubernetesZone] != podRemovalZone {
+			continue
+		}
+		if pod.Spec.NodeName == "" {
+			r.Log.Info(fmt.Sprintf("NodeName is empty for pod %s.", pod.Name))
+			continue
+		}
+		vhost := nodeIdToPodMap[pod.GetName()]
+		err = r.HumioClient.SetIsBeingEvicted(ctx, humioHttpClient, req, vhost, true)
+		if err != nil {
+			return r.logErrorAndReturn(err, fmt.Sprintf("failed to set eviction for vhost %d", vhost))
+		}
+		// wait for eviction status to be updated
+		isEvicted := false
+		for i := 0; i < waitForPodTimeoutSeconds && !isEvicted; i++ {
+			nodesStatus, err := r.getClusterNodesStatus(ctx, humioHttpClient, req)
+			if err != nil {
+				return err
+			}
+
+			for _, node := range nodesStatus {
+				if node.GetId() == vhost && *node.GetIsBeingEvicted() {
+					isEvicted = true
+					break
+				}
+			}
+
+			r.Log.Info(fmt.Sprintf("validating node data eviction is in progress for vhost %d", vhost))
+			time.Sleep(time.Second * 1)
+		}
+
+		if !isEvicted {
+			return r.logErrorAndReturn(err, fmt.Sprintf("failed to set eviction status for vhost %d", vhost))
+		}
+
+		pod.Annotations[kubernetes.PodMarkedForDataEviction] = "true"
+
+		// poll eviction status
+		pollTick := time.Tick(60 * time.Second)
+		for _ = range pollTick {
```

Comment:
I'm not sure this is really the way to go. If we do it like this, it means the operator would effectively be "locked up" until this loop finishes.
What we typically do is "block for at most X time" in a single`r.Reconcile()` execution. This way there's an upper bound on how long time the operator would be stuck in there.

The reason why I say this is that a single instance of humio-operator may be managing many CRD's, and so we can't have it's behavior to block for too long time because it would block reconciliation of all the other objects.

This is why we generally tend to follow this:
1. do something
2. optional: wait at most X
3. requeue
Then rely on requeue to pick up from where we left off.

Reply:
The mean idea behind "not doing too much" before requeuing is that the operator may be interrupted at any time, so we can't assume any updates in-memory will survive for long (or even short I guess since it can be restarted at any time).

Reply:
Fixed ❓ 😅